Initialize enterprise api client with provided user

zubair-arbi opened the pull request:
ENT-624
@asadiqbal08 @saleem-latif @douglashall 
Initialize and authenticate enterprise api client with the provided user or use enterprise service user by default.

douglashall commented on openedx/features/enterprise_support/api.py:
```diff
@@ -102,11 +97,12 @@ class EnterpriseApiClient(object):
     """
 
-    def __init__(self):
+    def __init__(self, user=None):
```
I think it would be better to create a new subclass of `EnterpriseApiClient` called `EnterpriseApiServiceClient` instead of defaulting to the service user here. See the [discovery clients](https://github.com/edx/edx-enterprise/blob/master/enterprise/api_client/discovery.py#L402) we have in edx-enterprise for an example.

Having separate classes makes it a little more obvious which type of user is going to be used to communicate with other services.

zubair-arbi commented:
@douglashall I have updated PR so that `request user` is used to initialize and authenticate API for `EnterpriseApiClient`and `ConsentApiClient`. Since we have decided to use `request user` instead of enterprise service user for accessing enterprise API so I have removed the optional `user` parameter for the API clients hence no need to have a subclass `EnterpriseApiServiceClient`. Let me know if this is right approach.

@haikuginger Please provide your feedback.

douglashall commented on openedx/features/enterprise_support/api.py:
```diff
@@ -458,7 +457,7 @@ def get_dashboard_consent_notification(request, user, course_enrollments):
                 break
 
-        client = ConsentApiClient()
+        client = ConsentApiClient(user=request.user)
```
@haikuginger Do we need to use the service user when making this request to the consent_required API endpoint or are we good with the request user?

haikuginger replied:
Doesn't matter. Request user should work.

douglashall commented on openedx/features/enterprise_support/api.py:
```diff
@@ -355,7 +354,7 @@ def consent_needed_for_course(request, user, course_id, enrollment_exists=False)
         consent_needed = False
     else:
-        client = ConsentApiClient()
+        client = ConsentApiClient(user=request.user)
```
@haikuginger Same question here. Are we good with using the request user?

haikuginger replied:
Request user is fine here.

douglashall commented on openedx/features/enterprise_support/api.py:
```diff
@@ -308,34 +303,38 @@ def enterprise_customer_for_request(request):
         # from the EnterpriseCustomer API.
         try:
-            ec = EnterpriseApiClient().get_enterprise_customer(ec_uuid)
+            enterprise_customer = EnterpriseApiClient(user=request.user).get_enterprise_customer(
```
I think you will need to use the service user here.

haikuginger replied:
This is actually where we do want to use the request user; this is the call that's being made when users enter courses.

douglashall replied:
Makes sense. I think we need to fix the [EnterpriseCustomerViewSet](https://github.com/edx/edx-enterprise/blob/master/enterprise/api/v1/views.py#L84) though and limit the GET results to those that the requesting user should have access to (i.e. the EnterpriseCustomer's that the user is linked to).

douglashall replied:
Turns out we did need to use the service user here because there are flows through this code that do not have an authenticated request user.

douglashall replied:
@zubair-arbi I think we need to completely refactor this function to use DB calls to look up the EnterpriseCustomer instead of API calls. We should create a `get_enterprise_customer(enterprise_customer_uuid)` function in this module to encapsulate those DB calls should we need to replace them with API calls somewhere down the line.

douglashall commented on common/djangoapps/enrollment/views.py:
```diff
@@ -598,8 +598,8 @@ def post(self, request):
             if (enterprise_course_consent or explicit_linked_enterprise) and has_api_key_permissions and enterprise_enabled():
-                enterprise_api_client = EnterpriseApiClient()
-                consent_client = ConsentApiClient()
+                enterprise_api_client = EnterpriseApiClient(user=request.user)
```
I think you need to use the service user here.

haikuginger replied:
Service user.

douglashall commented on common/djangoapps/enrollment/views.py:
```diff
@@ -598,8 +598,8 @@ def post(self, request):
-                enterprise_api_client = EnterpriseApiClient()
-                consent_client = ConsentApiClient()
+                enterprise_api_client = EnterpriseApiClient(user=request.user)
+                consent_client = ConsentApiClient(user=request.user)
```
@haikuginger Request user or service user?

haikuginger replied:
Service user.

zubair-arbi commented:
@douglashall @haikuginger I have addressed your feedback of using enterprise service user for `enterprise api client` and `consent api client` in enrollment view. Please have a look.

zubair-arbi commented:
@douglashall @haikuginger I am waiting for your review on this PR. Please have a look.

haikuginger commented on openedx/features/enterprise_support/api.py:
```diff
@@ -97,17 +92,31 @@ def consent_required(self, enrollment_exists=False, **kwargs):
+    Class for producing an Enterprise Consent API client with service user.
+    """
+
+    def __init__(self):
```
Do you want to abstract this into an `EnterpriseServiceClientMixin`? That way, we can just drop that mixin in front of the relevant base client (`ConsentApiClient` or `EnterpriseApiClient`) in the inheritance order and not even need to do anything in the class body.

zubair-arbi replied:
Your suggestion would work but I think explicitly getting `EnterpriseApiServiceClient` and `ConsentApiServiceClient` make them more readable.

haikuginger replied:
Right; I'm saying something like this:

```python
class EnterpriseApiServiceClient(EnterpriseServiceUserClient, EnterpriseApiClient):
    pass
```

zubair-arbi replied:
I have updated the PR with this suggestion. Please have a look.

haikuginger commented on openedx/features/enterprise_support/tests/test_api.py:
```diff
@@ -26,6 +30,9 @@
 from student.tests.factories import UserFactory
 
 
+ENTERPRISE_SERVICE_WORKER_USERNAME = 'enterprise_worker'
```
Should we declare this here or get it from `settings`?

zubair-arbi replied:
Thanks for pointing this out. I will use this value from settings.

haikuginger commented:
@zubair-arbi, looks good! My previous approval stands.

zubair-arbi commented:
Rebased from master and squash from commit db0263dec4dc93a429cf25d9ec9a3a48e7e2f776.
